### PR TITLE
Feature/awf/cleanup time agg param#212

### LINKED ISSUE #212: Modify selection of time aggregation parameter
Resolve `TODO` in `chart.component.ts`.

Either make `time_aggregation` query parameter user-selectable, or if always yearly, move to an `enum` or other form of constant.

See commentary on #182 and #210.

### PR DESCRIPTION
**Pull request by CloudNiner**
## Overview

Replaces the `string` type of the `time_aggregation` indicator param with a new enum type `TimeAggParam`. This allows for improved type checking.

Assigned both of you so you can see this change. @flibbertigibbet because I mentioned I was making it in a previous PR you authored, and @fungjj92 so it can inform your work on the percentile indicator params.

### Demo

No user facing changes

### Notes

To make this easier, I replaced the subclassing of IndicatoryQueryOpts by making IndicatorQueryOpts.params have a new IndicatorParams interface that can be extended instead of the QueryOpts interfaces. This provides a much improved type interface for these, and allows us to better type check the params as they're passed around the app, especially when we construct requests to IndicatorService and when we pass the params around via EventEmitters.

The models folder is getting pretty big, and its now very unwieldy to import one item from each file everywhere. I considered adding an index.ts to the models folder and exporting all of the items in that folder there and updating import statements across the app, but that would have cluttered the intent of this PR. Maybe a good separate issue.

## Testing Instructions

App should continue to function

Closes #212

### CODE REVIEW

**Comment by CloudNiner** [comment]
Most of below lifted from an offline discussion, but copied here for visibility (@jcahail):

I chose to go with only the `yearly` time agg option and not provide a user selectable dropdown. The quarterly and monthly options provide too much noise unless a user is zoomed in to a few year's time range, at which point it's not really possible to glean useful information from it.

Monthly option for all years:
screen shot 2017-09-01 at 10 04 56

Monthly option at ~5yr range:
screen shot 2017-09-01 at 10 06 32

We could potentially revisit this if we hear a bunch of users requesting it, but for now its less effort and less complicated for users to just omit the option.

File: src/app/charts/extra-params-components/threshold.component.ts [hunk 1]
```diff
@@ -62,11 +62,11 @@ export class ThresholdComponent implements AfterViewInit, OnInit {
     ngAfterViewInit() {
         // Since valueChanges triggers initially before parent is ready, wait until
         // parent is ready here and trigger it to draw chart with extra parameters.
-        this.thresholdParamSelected.emit({data: {
+        this.thresholdParamSelected.emit({
```
Comment (flibbertigibbet) [inline src/app/charts/extra-params-components/threshold.component.ts]: Why is the `data` object removed here? You had asked for it to be added in another PR.

File: src/app/models/indicator-params.model.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { ClimateModel } from './climate-model.model';
+import { TimeAggParam } from './time-agg-param.enum';
+
+export interface IndicatorParams {
+    climateModels?: ClimateModel[];
```
Comment (flibbertigibbet) [inline src/app/models/indicator-params.model.ts]: Just noticed these are all optional. Shouldn't they be required?

File: src/app/services/chart.service.ts [hunk 3]
```diff
@@ -17,7 +17,6 @@ export class ChartService {
 
     private timeOptions = {
           'yearly': '%Y',
-          'daily': '%Y-%m-%d',
```
Comment (flibbertigibbet) [inline src/app/services/chart.service.ts]: Should there be a quarterly format added here?

File: src/app/charts/chart.component.ts [hunk 4]
```diff
@@ -115,18 +116,15 @@ export class ChartComponent implements OnChanges, OnDestroy, AfterViewInit {
         this.cancelDataRequest();
     }
 
-    updateChart(extraParams: any) {
+    updateChart(extraParams: IndicatorParams) {
         this.cancelDataRequest();
         this.chartData = [];
         this.rawChartData = [];
```
Comment (fungjj92) [inline src/app/charts/chart.component.ts]: I'm not understanding why the typecheck doesn't complain when ThresholdIndicatorParams are sent through

File: src/app/models/indicator-params.model.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { ClimateModel } from './climate-model.model';
```
Comment (fungjj92) [inline src/app/models/indicator-params.model.ts]: Now there are two models:
indicator-parameters.model.ts
indicator-params.model.ts

in the models folder, and that is confusing

File: src/app/charts/chart.component.ts (hunk 4 above)
Comment (CloudNiner) [reply src/app/charts/chart.component.ts]: Because `interface ThresholdIndicatorParams` extends `interface `IndicatorParams`

File: src/app/services/chart.service.ts (hunk 3 above)
Comment (CloudNiner) [reply src/app/services/chart.service.ts]: Hmm, we don't use it but it wouldn't hurt to be thorough

File: src/app/models/indicator-params.model.ts (hunk 2 above)
Comment (CloudNiner) [reply src/app/models/indicator-params.model.ts]: No, because none of these params are required to make an API request, they all have defaults.

File: src/app/charts/extra-params-components/threshold.component.ts (hunk 1 above)
Comment (CloudNiner) [reply src/app/charts/extra-params-components/threshold.component.ts]: I think I had misunderstood what the original problem was based on the comments before, combined with forgetting the details of how `EventEmitter`s work. What's here now is the proper way to return data via event emitters, if you do not need the [DOM Event](https://developer.mozilla.org/en-US/docs/Web/API/Event)

File: src/app/charts/chart.component.ts (hunk 4 above)
Comment (fungjj92) [reply src/app/charts/chart.component.ts]: Ah interesting. Extending works the opposite direction I was expecting, then. Cool.

File: src/app/models/indicator-params.model.ts (hunk 5 above)
Comment (CloudNiner) [reply src/app/models/indicator-params.model.ts]: Any suggestions for better names? otherwise I'll try to come up with something

Comment (fungjj92) [reply src/app/models/indicator-params.model.ts]: We've been using the term "extra params"
Maybe 
indicator-extra-params?

Comment (fungjj92) [reply src/app/models/indicator-params.model.ts]: Or perhaps indicator-query-parameters

File: src/app/services/chart.service.ts (hunk 3 above)
Comment (CloudNiner) [reply src/app/services/chart.service.ts]: Actually, can't do this now. It would require additional refactoring. In the quarterly case, we need to convert the keys in the API (e.g. 2010-3) into the start date for a "quarter" which would be July 1, 2010 for 2010-3, and that can't be done with `D3.timeParse`. We'd need this refactor for pretty much all of the other remaining valid time aggs if we chose to display them in the lab UI.

File: src/app/models/indicator-params.model.ts (hunk 5 above)
Comment (CloudNiner) [reply src/app/models/indicator-params.model.ts]: We already have indicator-query-opts and these params aren't always the subset that are the extra params. They're really the GET params to an API request. What about the following refactor:
- `IndicatorQueryOpts -> IndicatorRequestOpts`
- `IndicatorParams -> IndicatorGetParams`
- Leave `IndicatorParam` unchanged

Comment (fungjj92) [reply src/app/models/indicator-params.model.ts]: OK IndicatorRequestOpts
#2 -- > IndicatorParamsList? or IndicatorQueryStringParams?
OK IndicatorParam

Comment (CloudNiner) [reply src/app/models/indicator-params.model.ts]: `IndicatorQueryParams` would just be a shorter version of your second option that isn't any less descriptive. How about that? Definitely like your second suggestion better than the first.
